### PR TITLE
Developer hub UI improvements

### PR DESCRIPTION
**Pull request by webdev51**
This PR adds the following to the developer hub applications page

- Added new sidebar for filtering applications in Developer hub
- Created pages for applications so they could be searched from the main search bar
- Tags behavior changed to be clickable with their own page
- Added tooltips to the AWS services icons

Screenshot:

image

CC: @tinyg210

### CODE REVIEW

File: content/en/references/configuration.md [hunk 1]
```diff
@@ -78,7 +78,7 @@ This section covers configuration values that are specific to certain AWS servic
 
 | Variable | Example Values | Description |
 | - | - | - |
-| `GRAPHQL_ENDPOINT_STRATEGY` | `legacy`\|`domain`\|`path` |  Governs how AppSync endpoints are created to access a GraphQL API (see [AppSync Endpoints]({{< ref "appsync#endpoints" >}})) |
```
Comment (webdev51) [inline content/en/references/configuration.md:81]: This is not related to the current PR. It is a fix for a build error which I was getting at this point saying "unambiguous reference" and I fixed it by giving the full path.

In developer hub, when I created an application with appsync tag, build started to fail and this error popped up

@lukqw

File: layouts/applications/list.html [hunk 2]
```diff
@@ -112,4 +122,133 @@ <h5 class="card-title d-flex align-items-center">
...
+        updateApplicationsUI();
+      });
+    };
+
+    var isFiltersApplied = function () {
+      let applied = false;
+      Object.keys(filtersState).forEach(function (key) {
+        if (filtersState[key].length !== 0) applied = true;
+      });
+      return applied;
```
Comment (lukqw) [inline layouts/applications/list.html:212]: seems like we could improve this by early exiting the function on line 210 with `return true` and replace line 212 with `return false`.

File: layouts/applications/list.html [hunk 3]
```diff
@@ -112,4 +122,133 @@ <h5 class="card-title d-flex align-items-center">
       margin: 0;
     }
   </style>
+  <script>
+    var $items = $(".entry");
+    var $noResults = $(".no-results");
+    var $clearFiltersBtn = $(".clear-filters-trigger");
+    var $showAllAppsBtn = $('.show-all-apps-btn')
```
[the remark this answers is not in the document]
Comment (webdev51) [reply layouts/applications/list.html:129]: Yes, that's a great idea.

For now, I changed quotes for now and put a semi-colon.

File: layouts/applications/list.html [hunk 4]
```diff
@@ -45,32 +44,37 @@ <h1>{{ .Title }}</h1>
...
                                               </a>
                                             </h5>
                                             <div>
                                               {{ range .services }}
-                                                <img src="/images/services/{{ . }}.svg" style="border-radius: 4px; max-width: 30px;margin-right: 5px">
+                                                <img src="/images/services/{{ . }}.svg" style="border-radius: 4px; max-width: 30px;margin-right: 5px; margin-top: 5px" data-toggle="tooltip" data-placement="top" title="{{index $servicesData .}}" aria-label="{{index $servicesData .}}">
                                               {{ end }}
                                               {{ with .pro }}
-                                              <img src="/images/pro-badge.png" style="border-radius: 4px; max-width: 30px;margin-right: 5px">
+                                              <img src="/images/pro-badge.png" style="border-radius: 4px; max-width: 30px;margin-right: 5px; margin-top: 5px"  data-toggle="tooltip" data-placement="top" title="Localstack Pro" aria-label="Localstack Pro">
```
[the remark this answers is not in the document]
Comment (webdev51) [reply layouts/applications/list.html:72]: Done :+1:

File: layouts/applications/list.html [hunk 5]
```diff
@@ -112,4 +122,133 @@ <h5 class="card-title d-flex align-items-center">
...
+      <div class="custom-control custom-checkbox">
+        <input type="checkbox" value="${prefix}-${item}" class="custom-control-input" id="${prefix}-${item}" aria-label="${prefix}-${item}">
+        <label class="custom-control-label" for="${prefix}-${item}">${lookup[item] || item}</label>
+      </div>
+      `;
+    };
+    var buildFiltersUI = function () {
+      let services = "",
+        deployments = "",
+        platforms = "";
```
[the remark this answers is not in the document]
Comment (webdev51) [reply layouts/applications/list.html:148]: Resolved

File: layouts/applications/list.html (hunk 2 above)
Comment (webdev51) [reply layouts/applications/list.html:212]: I did the exact same at first but unfortunately `.forEach` [can't be skipped from in between](https://stackoverflow.com/a/2641374/10228249) and I don't usually use for loops to iterate arrays so I went ahead with this solution. One option would be to use `.some` / `.every` to resolve it if you think this is an issue

File: layouts/applications/list.html [hunk 6]
```diff
@@ -112,4 +122,133 @@ <h5 class="card-title d-flex align-items-center">
...
+    var $clearFiltersBtn = $(".clear-filters-trigger");
+    var $showAllAppsBtn = $('.show-all-apps-btn')
+    var filtersState = {
+      services: [],
+      deployments: [],
+      platforms: [],
+      cloudpods: [],
+    };
+
+    var checkboxItemHtmlGenerator = function (prefix, lookup, item) {
```
[the remark this answers is not in the document]
Comment (webdev51) [reply layouts/applications/list.html:137]: Sure, changed

File: layouts/applications/list.html [hunk 7]
```diff
@@ -112,4 +132,136 @@ <h5 class="card-title d-flex align-items-center">
...
+      });
+
+      // Set removes duplicates
+      services = Array.from(new Set(services.slice(0, -1).split(",")));
+      platforms = Array.from(new Set(platforms.slice(0, -1).split(",")));
+      deployments = Array.from(new Set(deployments.slice(0, -1).split(",")));
+
+      let servicesHtml = "",
+        deploymentsHtml = "",
+        platformsHtml = "";
```
Comment (lukqw) [inline layouts/applications/list.html:180]: please change to single line let statements

File: layouts/applications/list.html [hunk 8]
```diff
@@ -112,4 +132,136 @@ <h5 class="card-title d-flex align-items-center">
...
+    };
+    var buildFiltersUI = function () {
+      let services = "";
+      let deployments = "";
+      let platforms = "";
+
+      $items.each(function (_i, value) {
+        let service = value.getAttribute("data-services"),
+          platform = value.getAttribute("data-platforms"),
+          deployment = value.getAttribute("data-deployments");
```
Comment (lukqw) [inline layouts/applications/list.html:166]: please change to single line let statements

File: layouts/applications/list.html (hunk 2 above)
Comment (lukqw) [reply layouts/applications/list.html:225]: interesting, today I learned!

**Review by lukqw — APPROVED** [review]
last remarks after which we can take this one live 🚀

File: layouts/applications/list.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukqw) [reply layouts/applications/list.html]: Thanks a lot!

File: layouts/applications/single.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<!doctype html>
+<html itemscope itemtype="http://schema.org/WebPage" lang="{{ .Site.Language.Lang }}" class="no-js">
+  <head>
+    <meta http-equiv="refresh" content="0; url='/applications/?id={{.Title | urlize}}'" />
+    {{ partial "head.html" . }}
```
[the remark this answers is not in the document]
Comment (webdev51) [reply layouts/applications/single.html:5]: Because even if this page is just redirecting, but as this page will be there in google searches, so having `meta` and SEO tags might a good idea as if someone shares this page on any social media, we should be providing meta tags for that. These are my thoughts, what do you think?

[the remark this answers is not in the document]
Comment (lukqw) [reply layouts/applications/single.html:5]: Sounds good

File: content/en/applications/create-applications.sh [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  | tr -d "'" \
+  | sed -E 's/[^a-zA-Z0-9]+/-/g' \
+  | sed -E 's/^-+|-+$//g' \
+  | tr "[:upper:]" "[:lower:]"
+}
+if [[ $# -eq 0 ]] ; then
+    echo 'Please pass path argument'
+    exit 0
+fi
+APPLICATIONS=$(jq '.applications' $1)
```
[the remark this answers is not in the document]
Comment (webdev51) [reply content/en/applications/create-applications.sh:22]: Added

File: content/en/applications/create-applications.sh [hunk 11]
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
+
+# Example usage
+# bash create-applications.sh ../../../data/developerhub/applications.json
+
+function slugify() {
```
[the remark this answers is not in the document]
Comment (webdev51) [reply content/en/applications/create-applications.sh:11]: Added

File: content/en/applications/create-applications.sh [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  | sed -E 's/[^a-zA-Z0-9]+/-/g' \
+  | sed -E 's/^-+|-+$//g' \
+  | tr "[:upper:]" "[:lower:]"
+}
+if [[ $# -eq 0 ]] ; then
+    echo 'Please pass path argument'
+    exit 0
+fi
+APPLICATIONS=$(jq '.applications' $1)
+for row in $(echo "${APPLICATIONS}" | jq -r '.[] | @base64'); do
```
[the remark this answers is not in the document]
Comment (webdev51) [reply content/en/applications/create-applications.sh:29]: Added

**Review by lukqw — APPROVED** [review]
Awesome set of changes! Looking forward to see this live! 🚀

**Review by lukqw — APPROVED** [review]
image
